server/src/main/java/com/microsoft/java/bs/core/internal/reporter/TestReportReporter.java

```diff
@@ -0,0 +1,147 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+package com.microsoft.java.bs.core.internal.reporter;
+
+import ch.epfl.scala.bsp4j.TaskId;
+import ch.epfl.scala.bsp4j.TaskStartParams;
+import ch.epfl.scala.bsp4j.TestFinish;
+import ch.epfl.scala.bsp4j.TestStart;
+import ch.epfl.scala.bsp4j.TestStatus;
+import org.gradle.tooling.events.FinishEvent;
+import org.gradle.tooling.events.OperationResult;
+import org.gradle.tooling.events.ProgressEvent;
+import org.gradle.tooling.events.StartEvent;
+import org.gradle.tooling.events.test.JvmTestOperationDescriptor;
+import org.gradle.tooling.events.test.TestFailureResult;
+import org.gradle.tooling.events.test.TestSkippedResult;
+import org.gradle.tooling.events.test.TestSuccessResult;
+
+import ch.epfl.scala.bsp4j.BuildClient;
+import ch.epfl.scala.bsp4j.BuildTargetIdentifier;
+import ch.epfl.scala.bsp4j.StatusCode;
+import ch.epfl.scala.bsp4j.TaskFinishParams;
+import ch.epfl.scala.bsp4j.TestReport;
+import org.gradle.tooling.internal.consumer.DefaultTestAssertionFailure;
+
+/**
+ * Implements {@link ProgressReporter} to record test results.
+ * Test summary report. e.g. number of fails.
+ */
+public class TestReportReporter extends ProgressReporter {
+
+  private final BuildTargetIdentifier btId;
+  private int successCount;
+  private int skippedCount;
+  private int failureCount;
+  private String exception;
+  private long testDuration;
+
+  /**
+   * initialise.
+   *
+   * @param btId the build target being tested.
+   */
+  public TestReportReporter(BuildTargetIdentifier btId, BuildClient client, String originId) {
+    super(client, originId);
+    this.btId = btId;
+    successCount = 0;
+    skippedCount = 0;
+    failureCount = 0;
+    exception = null;
+    testDuration = 0;
+  }
+
+  @Override
+  public void statusChanged(ProgressEvent event) {
+    if (client != null) {
+      if (event.getDescriptor() instanceof JvmTestOperationDescriptor) {
+        // only report on methods
+        JvmTestOperationDescriptor descriptor = (JvmTestOperationDescriptor) event.getDescriptor();
+        if (descriptor.getClassName() != null && descriptor.getMethodName() != null) {
+          TaskId taskId = getTaskId(descriptor.getDisplayName());
+          if (event instanceof StartEvent) {
+            TaskStartParams startParam = new TaskStartParams(taskId);
+            startParam.setMessage("Start test");
+            startParam.setDataKind("test-start");
+            startParam.setEventTime(event.getEventTime());
+            TestStart testStart = new TestStart(event.getDisplayName());
+            // TODO Gradle does not provide file/position data
+            startParam.setData(testStart);
+            client.onBuildTaskStart(startParam);
+          } else if (event instanceof FinishEvent) {
```

Comment:
nit: the enhanced instanceof pattern introduced in Java 14 may simplify the code just a little bit.

Reply:
I agree - do the VSCode java tools all require jdk 17 now?  I wasn't sure if 11 is still supported (to run under)

Reply:
Yes, in VS Code Java, the build server will be launched using jre 17 now.